File: grind75/week2/004-Merge-Intervals-Medium/004-Merge-Intervals-Medium.cpp
// Problem link: https://leetcode.com/problems/merge-intervals/description/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {

        std::sort(intervals.begin(), intervals.end());
        vector<vector<int>> ans;

        int start = 0;
        int end = 1;
        int prevStart = intervals[0][start];
        int prevEnd = intervals[0][end];

        for(int i = 1; i < intervals.size(); i++)
        {
            if (intervals[i][start] <= prevEnd)
            {
                prevEnd = std::max(prevEnd, intervals[i][end]);
            }
            else
            {
                ans.push_back({prevStart, prevEnd});
                prevStart = intervals[i][start];
                prevEnd = intervals[i][end];
            }

        }

        // Either we merge and don't append quite yet
        // or append but set the last one and not append it 
        //  in either case, we have an interval to append
        ans.push_back({prevStart, prevEnd});

        return ans;
        
    }
};

File: grind75/week2/004-Merge-Intervals-Medium/004-Merge-Intervals-Medium.py
# problem link: https://leetcode.com/problems/merge-intervals/description/

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:

        # output is a list of list

        intervals.sort()

        start = 0
        end = 1
        ans = []
        prev = intervals[0]
        for interval in intervals[1:]:

            if interval[start] <= prev[end]:
                prev[end] = max(interval[end], prev[end])
            else:
                ans.append(prev)
                prev = interval
        
        # Either we merge and don't append quite yet
        # or append but set the last one and not append it 
        # in either case, we have an interval to append
        ans.append(prev)

        
        return ans

